File: guests/haskell/typeface.hs
{-# LANGUAGE OverloadedStrings #-}

-- The typeface scene, Haskell port — guests/rust/typeface.rs,
-- tools/scenes/typeface.steps.

import Data.IORef (newIORef, readIORef, writeIORef)

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  draftRef <- newIORef ("" :: Text)
  -- Out here: Build is a pure state monad, so the one IO the brand call needs
  -- happens before the transaction opens.
  font <- asset "fonts/sora-wght.ttf"
  buildTx app $ do
    -- BEFORE THE FIRST MOUNT, per the set-once wall.
    brandTypeface "Sora" [TFontAsset font]
    window primary [WTitle "typeface", WSize 480 360]

    heading <- signal (T.pack "typeface")
    status <- signal (T.pack "ready")

    root <-
      column
        []
        [ -- The heading's text style OVERRIDES the root font, so this is
          -- the label a root-only lowering leaves in the system face.
          labelBound heading [Role Heading, A11yId ("title" :: Text)], -- label#0
          labelBound status, -- label#1
          -- A field AND a textarea, because the swap reaches them by
          -- DIFFERENT routes.
          entryOn (writeIORef draftRef), -- entry#0
          textarea, -- textarea#0
          buttonOn -- button#0
            "Go"
            ( do
                draft <- readIORef draftRef
                submitTx app (writeSignal status ("clicked " <> draft))
            )
        ]
    mount root
  assetClose font

File: guests/rust/typeface.rs
//! The typeface conformance scene (tools/scenes/typeface.steps): NAME NO
//! SIZE ANYWHERE, and the font is VENDORED so no fallback can pass.

use kaya::Occurrence;

pub(crate) fn app(ctx: kaya::AppCtx) {
    let (status, field, go) = ctx.apply(|tx| {
        // BEFORE THE FIRST MOUNT, per the set-once wall.
        let font = tx.asset("fonts/sora-wght.ttf");
        tx.brand_typeface_with("Sora", &[], Some(&font));
        tx.window(kaya::DEFAULT_WINDOW).title("typeface").size(480.0, 360.0);
        let heading = tx.signal("typeface");
        let status = tx.signal("ready");
        let (root, (field, go)) = tx
            .column(|tx| {
                // A heading OVERRIDES the root font: a root-only lowering shows.
                tx.label(heading).role(kaya::Role::Heading).a11y_id("title"); // label#0
                tx.label(status); // label#1
                // Two DIFFERENT routes: one alone cannot see a half swap.
                let field = tx.entry().id(); // entry#0
                tx.textarea(); // textarea#0
                let go = tx.button("Go").id(); // button#0
                (field, go)
            })
            .into_parts();
        tx.mount(root);
        (status, field, go)
    });

    let mut draft = String::new();
    loop {
        match ctx.next() {
            Occurrence::TextChanged { id, text } if id == field => draft = text,
            Occurrence::ButtonClicked { id } if id == go => {
                let text = draft.clone();
                ctx.apply(|tx| tx.write(status, format!("clicked {text}")));
            }
            _ => {}
        }
    }
}

fn main() {
    kaya::run(app)
}
